contracts/MechMarketplace.sol

```diff
@@ -381,132 +376,268 @@ contract MechMarketplace is IErrorsMarketplace {
+            if (priorityMech == address(0)) {
+                revert ZeroAddress();
+            }
 
-        // Check for request existence
-        if (priorityMech == address(0)) {
-            revert ZeroAddress();
-        }
+            // Check payment type
+            if (requestInfo.paymentType != paymentType) {
+                revert WrongPaymentType(paymentType);
+            }
 
-        // Check if request has been delivered
-        if (mechDelivery.deliveryMech != address(0)) {
-            revert AlreadyDelivered(requestId);
-        }
+            // Check if request has been delivered
+            if (requestInfo.deliveryMech != address(0)) {
+                continue;
+            }
 
-        // If delivery mech is different from the priority one
-        if (priorityMech != msg.sender) {
-            // Within the defined response time only a chosen priority mech is able to deliver
-            if (block.timestamp > mechDelivery.responseTimeout) {
-                // Decrease priority mech karma as the mech did not deliver
-                IKarma(karma).changeMechKarma(priorityMech, -1);
-                // Revoke request from the priority mech
-                IMech(priorityMech).revokeRequest(requestId);
-            } else {
-                // Priority mech responseTimeout is still >= block.timestamp
-                revert PriorityMechResponseTimeout(mechDelivery.responseTimeout, block.timestamp);
+            // TODO continue instead of revert? This would let us just ignore incorrect requests without reverting the bulk
+            // TODO Additional thought: The price between posting requests and delivering could have changed, thus just continue such that mech cleans those request info on their side
+            // Check for actual mech delivery rate
+            requesterDeliveryRates[i] = requestInfo.deliveryRate;
+            if (deliveryRates[i] > requesterDeliveryRates[i]) {
+                revert Overflow(deliveryRates[i], requestInfo.deliveryRate);
             }
-        }
 
-        // Record the actual delivery mech
-        mechDelivery.deliveryMech = msg.sender;
+            // If delivery mech is different from the priority one
+            if (priorityMech != msg.sender) {
+                // Within the defined response time only a chosen priority mech is able to deliver
+                if (block.timestamp > requestInfo.responseTimeout) {
+                    // Decrease priority mech karma as the mech did not deliver
+                    // Needs to stay atomic as each different priorityMech can be any address
+                    IKarma(karma).changeMechKarma(priorityMech, -1);
+                } else {
+                    // TODO continue instead of revert? This would let us just ignore incorrect requests without reverting the bulk
+                    // TODO Additional thought: revert as mech has just miscalculated time, does not mean it has to clear that request
```

Comment:
scenarios needed

Reply:
Delivery mech is trying to deliver a request during the time when priority mech still has its priority time to deliver.